Datepicker: Add isInvalidDay support

psealock opened the pull request:
## Description
Add the concept of unselectable days to the DatePicker's calendar by exposing `react-dates` prop [`isOutsideRange`](https://github.com/airbnb/react-dates/blob/b002152c6229b2e9b0c5c2d80a15469015ada680/src/components/DayPickerSingleDateController.jsx#L43). Here is an example with Mondays made as invalid selections:

![screen shot 2018-12-18 at 12 52 13 pm](https://user-images.githubusercontent.com/1922453/50123136-6d2caa00-02c4-11e9-9a18-f237d6d6b5ac.png)

## What about invalidating dates via the inputs?

The changes here affect only the calendar (`<DatePicker />` component) and not the inputs (`<TimePicker />`). Ideally, the same validation is passed along to the inputs to prevent selection of invalid days via the keyboard. That will take some design work to surface why an invalidation has occurred. I propose to handle that separately.

## Testing instructions

Since the concept of invalid days has no meaning for the text editor, changing code temporarily is the only way to evaluate this change.

1. Add this prop to invalidate Mondays
```js
isInvalidDate={ ( date ) => 1 === date.getDay() }
```
to the implementation of `<DatePicker />`, https://github.com/WordPress/gutenberg/blob/0c734e1867b50a9ee1eb19a08d6b740d66c5f2ad/packages/components/src/date-time/index.js#L48-L51

2. Go to a new post `/wp-admin/post-new.php#/`.
3. Open the date picker.
4. See Mondays are not selectable via the calendar.
5. Using VoiceOver, confirm focus of an invalid calendar date adds a "Not available" to the description.

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->
New feature: Expose a prop, `isInvalidDay `, to `<DatePicker />`.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://wordpress.org/gutenberg/handbook/designers-developers/ -->

psealock committed:
Datepicker: Add inValidDay support

timmyc commented on packages/components/src/date-time/date.js:
```diff
@@ -36,7 +36,7 @@ class DatePicker extends Component {
 
 	render() {
-		const { currentDate } = this.props;
+		const { currentDate, isInvalidDay } = this.props;
```
Probably not totally required but should we provide a default prop for this of `false` just to be extra explicit?

psealock replied:
Thanks @timmyc !

I could set a default prop or just `{ isInvalidDay = false } = this.props`, but I think its best to pass along an `undefined` value to react-dates's own default prop for `isOutsideRange`:

https://github.com/airbnb/react-dates/blob/d572bea9029739089bd200026be3aad7684a8b02/src/components/DayPickerSingleDateController.jsx#L100

The reason is because React handles an `undefined` prop differently than `null` or possibly `false` (https://reactjs.org/docs/react-component.html#defaultprops).

timmyc reviewed (commented):
Looking good to me, just one minor comment.

timmyc reviewed (approved):
This is looking good to me.

aduth reviewed (changes requested):
The DateTime component has a README.md . The prop should be documented.

psealock commented:
Thanks for the review @aduth. Changes made to DateTime Readme.

aduth commented:
Moment is not otherwise exposed as part of the public interface, and I don't think we'd want to start now (related: #11418).

What would you think about passing the `Date` object instead? 

https://momentjs.com/docs/#/displaying/as-javascript-date/

psealock commented:
@aduth I've modified how `isInvalidDay` is passed on to `react-dates` such that the prop function's argument is a `Date` object instead of Moment.

I updated the test instructions on this PR. Thanks again for the input.

aduth commented on packages/components/src/date-time/README.md:
```diff
@@ -60,3 +60,10 @@ Whether we use a 12-hour clock. With a 12-hour clock, an AM/PM widget is display
+
+### isInvalidDay
+
+A function with a Date object representing a day as an argument returning a Boolean to signify if the day is valid or not. This function is passed to react-dates' `isOutsideRange` prop.
```
>This function is passed to react-dates' `isOutsideRange` prop.

But:

1. It's not the function we pass. We pass a wrapped function which handles the conversion from Moment to Date in the callback.
2. This component is an abstraction. The consumer needn't and shouldn't know that `react-dates` exists, largely because we don't want to make the commitment that it will always be the underlying implementation. Otherwise we put ourselves in a bind for future maintenance if/when we decide to go a different direction.

```suggestion
A callback function which receives a Date object representing a day as an argument, and should return a Boolean to signify if the day is valid or not.
```

aduth commented on packages/components/src/date-time/date.js:
```diff
@@ -36,7 +36,7 @@ class DatePicker extends Component {
 
 	render() {
-		const { currentDate } = this.props;
+		const { currentDate, isInvalidDay } = this.props;
```
Sorry, I'm nitpicking at this point, but the term "Day" is ambiguous to me, where the value itself to check is a `Date` object. I'm also a bit wary about the negation here, regardless of what `react-dates` is doing, since it seems a bit unnecessary to reverse what is essentially a validity check.

Thoughts on `isValidDate` instead?

There's a further argument about the naming of the prop implying that its value is itself a boolean, when in fact it's a callback function. I'm not sure what precedent there is here for these sorts of callback functions. I'm not overly concerned about it.

psealock replied:
>I'm also a bit wary about the negation here, regardless of what react-dates is doing, since it seems a bit unnecessary to reverse what is essentially a validity check.

One advantage of the negation relates to the optionality of the prop. By default, all dates are valid. It would seem odd to me that by adding a positively-framed check, `isValidDate`, I'd have to explicitly return true to enable a date that is otherwise enabled by default.

>the term "Day" is ambiguous to me

Great point. 

My slight preference is `isInvalidDate`, but I can go with the positive `isValidDate` as well.

aduth replied:
I'd be fine with `isInvalidDate`

psealock replied:
Cool, done and rebased.

aduth reviewed (approved):
Looks great 👍 Thanks for your patience.
